File: src/portfolio/black_litterman.rs
//! Portfolio Optimization - Black-Litterman Model
//! 
//! Implements a simplified Black-Litterman model combining market equilibrium
//! with proprietary alpha views. Calculates optimal multi-asset weight allocations
//! (BTC, ETH, SOL) balancing expected returns against the covariance matrix.

use std::collections::HashMap;
use nalgebra::{Matrix2, Matrix3, MatrixXx, Vector2, Vector3, VectorX};

use tracing::{debug, info, warn};

/// Black-Litterman configuration
#[derive(Debug, Clone)]
pub struct BlackLittermanConfig {
    /// Risk aversion coefficient (lambda)
    pub risk_aversion: f64,
    /// Scaling factor for uncertainty in views (tau)
    pub tau: f64,
    /// Market capitalization weights
    pub market_caps: HashMap<String, f64>,
    /// Risk-free rate (annualized)
    pub risk_free_rate: f64,
}

impl Default for BlackLittermanConfig {
    fn default() -> Self {
        let mut market_caps = HashMap::new();
        // Approximate crypto market caps (would be updated dynamically)
        market_caps.insert("BTC".to_string(), 0.50);
        market_caps.insert("ETH".to_string(), 0.30);
        market_caps.insert("SOL".to_string(), 0.20);

        Self {
            risk_aversion: 2.5,
            tau: 0.05,
            market_caps,
            risk_free_rate: 0.05, // 5% annual
        }
    }
}

/// Alpha view from proprietary models
#[derive(Debug, Clone)]
pub struct AlphaView {
    pub asset: String,
    pub expected_return: f64,
    pub confidence: f64, // 0.0 to 1.0
    pub view_type: ViewType,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ViewType {
    Absolute,   // Expected return for single asset
    Relative,   // Asset A will outperform Asset B by X%
}

/// Black-Litterman result
#[derive(Debug, Clone)]
pub struct BlackLittermanResult {
    pub optimal_weights: HashMap<String, f64>,
    pub expected_returns: HashMap<String, f64>,
    pub portfolio_expected_return: f64,
    pub portfolio_volatility: f64,
    pub sharpe_ratio: f64,
}

/// Black-Litterman optimizer
pub struct BlackLittermanOptimizer {
    config: BlackLittermanConfig,
    covariance_matrix: HashMap<(String, String), f64>,
    assets: Vec<String>,
}

impl BlackLittermanOptimizer {
    /// Create a new Black-Litterman optimizer
    pub fn new(config: BlackLittermanConfig) -> Self {
        let assets: Vec<String> = config.market_caps.keys().cloned().collect();
        
        Self {
            config,
            covariance_matrix: HashMap::new(),
            assets,
        }
    }

    /// Set covariance matrix from historical data
    pub fn set_covariance(&mut self, covariances: HashMap<(String, String), f64>) {
        self.covariance_matrix = covariances;
    }

    /// Calculate implied equilibrium returns from market cap weights
    pub fn calculate_equilibrium_returns(&self) -> HashMap<String, f64> {
        let lambda = self.config.risk_aversion;
        let mut returns = HashMap::new();

        // Σ * w gives the marginal contribution to variance
        // Π = λ * Σ * w (implied excess returns)
        
        for asset in &self.assets {
            let weight = self.config.market_caps.get(asset).copied().unwrap_or(0.0);
            
            // Calculate marginal variance contribution
            let mut marginal_var = 0.0;
            for other_asset in &self.assets {
                let other_weight = self.config.market_caps.get(other_asset).copied().unwrap_or(0.0);
                let cov = self.covariance_matrix.get(&(asset.clone(), other_asset.clone()))
                    .copied()
                    .or_else(|| self.covariance_matrix.get(&(other_asset.clone(), asset.clone())).copied())
                    .unwrap_or(0.0);
                
                marginal_var += cov * other_weight;
            }

            // Implied return = λ * marginal variance contribution
            let implied_return = lambda * marginal_var;
            returns.insert(asset.clone(), implied_return + self.config.risk_free_rate);
        }

        returns
    }

    /// Combine equilibrium with views using Black-Litterman formula
    pub fn calculate_optimal_weights(
        &self,
        views: &[AlphaView],
    ) -> BlackLittermanResult {
        let n = self.assets.len();
        
        if n == 0 {
            return self.empty_result();
        }

        // Build covariance matrix
        let sigma = self.build_covariance_matrix(n);
        
        // Get equilibrium returns
        let pi = self.calculate_equilibrium_returns();
        
        // If no views, return equilibrium weights
        if views.is_empty() {
            return self.result_from_weights(self.config.market_caps.clone(), &pi);
        }

        // Build view matrix P and view vector Q
        let k = views.len();
        let mut p_matrix = vec![vec![0.0; n]; k];
        let mut q_vector = vec![0.0; k];
        let mut omega_diag = vec![0.0; k]; // Uncertainty in views

        for (i, view) in views.iter().enumerate() {
            if let Some(asset_idx) = self.assets.iter().position(|a| a == &view.asset) {
                match view.view_type {
                    ViewType::Absolute => {
                        p_matrix[i][asset_idx] = 1.0;
                        q_vector[i] = view.expected_return;
                        // Omega = (confidence * τ * σ²)^-1
                        let var = sigma[asset_idx][asset_idx];
                        omega_diag[i] = (view.confidence * self.config.tau * var).max(1e-10);
                    }
                    ViewType::Relative => {
                        // Would need second asset for relative views
                        p_matrix[i][asset_idx] = 1.0;
                        q_vector[i] = view.expected_return;
                        omega_diag[i] = self.config.tau * 0.1; // Simplified
                    }
                }
            }
        }

        // Black-Litterman formula:
        // E[R] = [(τΣ)^-1 + P'Ω^-1P]^-1 * [(τΣ)^-1Π + P'Ω^-1Q]
        
        // For simplicity, use scalar approximation
        let tau_sigma_inv = self.matrix_scalar_multiply(&sigma, 1.0 / self.config.tau);
        
        // Combine views with equilibrium
        let blended_returns = self.blend_views_with_equilibrium(&pi, views, n);

        // Calculate optimal weights using mean-variance optimization
        // w* = (λΣ)^-1 * E[R]
        let lambda = self.config.risk_aversion;
        let optimal_weights = self.calculate_mean_variance_weights(&blended_returns, lambda);

        // Normalize weights to sum to 1
        let weight_sum: f64 = optimal_weights.values().sum();
        let mut normalized_weights = optimal_weights.clone();
        if weight_sum > 0.0 {
            for w in normalized_weights.values_mut() {
                *w /= weight_sum;
            }
        }

        self.result_from_weights(normalized_weights, &blended_returns)
    }

    /// Blend views with equilibrium returns
    fn blend_views_with_equilibrium(
        &self,
        equilibrium: &HashMap<String, f64>,
        views: &[AlphaView],
        n: usize,
    ) -> HashMap<String, f64> {
        let mut blended = equilibrium.clone();
        
        // Tau controls how much we deviate from equilibrium
        let tau = self.config.tau;

        for view in views {
            if let Some(current) = blended.get_mut(&view.asset) {
                // Weighted average of equilibrium and view
                let view_weight = view.confidence * (1.0 - tau);
                let equilibrium_weight = 1.0 - view_weight;
                
                *current = equilibrium_weight * (*current) + view_weight * view.expected_return;
            } else {
                blended.insert(view.asset.clone(), view.expected_return);
            }
        }

        blended
    }

    /// Calculate mean-variance optimal weights
    fn calculate_mean_variance_weights(
        &self,
        expected_returns: &HashMap<String, f64>,
        lambda: f64,
    ) -> HashMap<String, f64> {
        let mut weights = HashMap::new();

        for asset in &self.assets {
            let ret = expected_returns.get(asset).copied().unwrap_or(0.0);
            let var = self.covariance_matrix.get(&(asset.clone(), asset.clone()))
                .copied()
                .unwrap_or(0.04); // Default 20% vol squared

            // Simple approximation: w_i = E[R_i] / (λ * σ²_i)
            let weight = if var > 1e-10 {
                ret / (lambda * var)
            } else {
                0.0
            };

            weights.insert(asset.clone(), weight.max(0.0)); // Long only
        }

        weights
    }

    /// Build NxN covariance matrix
    fn build_covariance_matrix(&self, n: usize) -> Vec<Vec<f64>> {
        let mut matrix = vec![vec![0.0; n]; n];

        for (i, asset_i) in self.assets.iter().enumerate() {
            for (j, asset_j) in self.assets.iter().enumerate() {
                let cov = self.covariance_matrix.get(&(asset_i.clone(), asset_j.clone()))
                    .copied()
                    .or_else(|| self.covariance_matrix.get(&(asset_j.clone(), asset_i.clone())).copied())
                    .unwrap_or_else(|| {
                        // Default: use variance on diagonal, 0 otherwise
                        if i == j {
                            0.04 // 20% annualized volatility
                        } else {
                            0.0
                        }
                    });
                
                matrix[i][j] = cov;
            }
        }

        matrix
    }

    /// Multiply matrix by scalar
    fn matrix_scalar_multiply(&self, matrix: &[Vec<f64>], scalar: f64) -> Vec<Vec<f64>> {
        matrix.iter()
            .map(|row| row.iter().map(|&v| v * scalar).collect())
            .collect()
    }

    /// Create result from weights and returns
    fn result_from_weights(
        &self,
        weights: HashMap<String, f64>,
        returns: &HashMap<String, f64>,
    ) -> BlackLittermanResult {
        // Calculate portfolio expected return
        let portfolio_return: f64 = weights.iter()
            .map(|(asset, weight)| {
                let ret = returns.get(asset).copied().unwrap_or(0.0);
                weight * ret
            })
            .sum();

        // Calculate portfolio volatility
        let mut portfolio_var = 0.0;
        for (asset_i, weight_i) in &weights {
            for (asset_j, weight_j) in &weights {
                let cov = self.covariance_matrix.get(&(asset_i.clone(), asset_j.clone()))
                    .copied()
                    .or_else(|| self.covariance_matrix.get(&(asset_j.clone(), asset_i.clone())).copied())
                    .unwrap_or(0.0);
                
                portfolio_var += weight_i * weight_j * cov;
            }
        }
        let portfolio_vol = portfolio_var.sqrt();

        // Calculate Sharpe ratio
        let excess_return = portfolio_return - self.config.risk_free_rate;
        let sharpe = if portfolio_vol > 1e-10 {
            excess_return / portfolio_vol
        } else {
            0.0
        };

        BlackLittermanResult {
            optimal_weights: weights,
            expected_returns: returns.clone(),
            portfolio_expected_return: portfolio_return,
            portfolio_volatility: portfolio_vol,
            sharpe_ratio: sharpe,
        }
    }

    fn empty_result(&self) -> BlackLittermanResult {
        BlackLittermanResult {
            optimal_weights: HashMap::new(),
            expected_returns: HashMap::new(),
            portfolio_expected_return: 0.0,
            portfolio_volatility: 0.0,
            sharpe_ratio: 0.0,
        }
    }

    /// Update market cap weights dynamically
    pub fn update_market_caps(&mut self, market_caps: HashMap<String, f64>) {
        // Normalize to sum to 1
        let total: f64 = market_caps.values().sum();
        if total > 0.0 {
            self.config.market_caps = market_caps.into_iter()
                .map(|(k, v)| (k, v / total))
                .collect();
            
            // Update assets list
            self.assets = self.config.market_caps.keys().cloned().collect();
        }
    }
}

/// Risk parity allocator as alternative
pub struct RiskParityAllocator {
    assets: Vec<String>,
    target_risk: f64,
}

impl RiskParityAllocator {
    pub fn new(assets: Vec<String>, target_risk: f64) -> Self {
        Self { assets, target_risk }
    }

    /// Calculate risk parity weights
    pub fn calculate_weights(&self, volatilities: &HashMap<String, f64>) -> HashMap<String, f64> {
        let mut weights = HashMap::new();
        let mut total_inverse_vol = 0.0;

        // Calculate inverse volatility weights
        for asset in &self.assets {
            let vol = volatilities.get(asset).copied().unwrap_or(0.2);
            let inverse_vol = 1.0 / vol.max(0.01);
            total_inverse_vol += inverse_vol;
            weights.insert(asset.clone(), inverse_vol);
        }

        // Normalize to sum to 1
        for weight in weights.values_mut() {
            *weight /= total_inverse_vol;
        }

        weights
    }
}

/// Print allocation summary
pub fn print_allocation_summary(result: &BlackLittermanResult) {
    println!("=== Black-Litterman Allocation ===");
    println!();
    println!("Optimal Weights:");
    for (asset, weight) in &result.optimal_weights {
        println!("  {}: {:.1}%", asset, weight * 100.0);
    }
    println!();
    println!("Expected Returns:");
    for (asset, ret) in &result.expected_returns {
        println!("  {}: {:.2}% annual", asset, ret * 100.0);
    }
    println!();
    println!("Portfolio Metrics:");
    println!("  Expected Return: {:.2}%", result.portfolio_expected_return * 100.0);
    println!("  Volatility: {:.2}%", result.portfolio_volatility * 100.0);
    println!("  Sharpe Ratio: {:.3}", result.sharpe_ratio);
}

File: python/portfolio/black_litterman.py
"""
Black-Litterman Optimizer integrating ML ensemble alpha views with market equilibrium.
Dynamically shifts multi-asset weights based on confidence scores from SOUL.md.
Memory-bounded implementation for 3GB RAM limit - no Pandas in hot-path.
"""

import numpy as np
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass


@dataclass
class InvestorView:
    """Represents a single investor view on asset returns."""
    assets: List[str]  # Assets involved in the view
    expected_return: float  # Expected excess return
    confidence: float  # Confidence level (0 to 1)
    view_type: str = "absolute"  # "absolute" or "relative"


class BlackLittermanOptimizer:
    """
    Black-Litterman portfolio optimizer combining market equilibrium with ML views.
    
    The model starts with market-implied equilibrium returns and adjusts them
    based on investor views weighted by their confidence levels.
    """
    
    def __init__(self, asset_ids: List[str], risk_aversion: float = 2.5,
                 tau: float = 0.05):
        """
        Initialize the Black-Litterman optimizer.
        
        Args:
            asset_ids: List of asset identifiers
            risk_aversion: Risk aversion coefficient (default 2.5)
            tau: Uncertainty scalar for prior (default 0.05)
        """
        self.asset_ids = asset_ids
        self.n_assets = len(asset_ids)
        self.risk_aversion = risk_aversion
        self.tau = tau
        
        # Cache for equilibrium returns
        self._equilibrium_returns: Optional[np.ndarray] = None
        self._last_market_caps_hash: int = 0
    
    def _compute_equilibrium_returns(self, market_caps: np.ndarray,
                                      cov_matrix: np.ndarray) -> np.ndarray:
        """
        Compute implied equilibrium returns from market capitalizations.
        
        Uses reverse optimization: given market weights and covariance,
        find the returns that would make these weights optimal.
        """
        # Market weights from capitalizations
        total_cap = np.sum(market_caps)
        market_weights = market_caps / (total_cap + 1e-10)
        
        # Implied equilibrium returns: Pi = delta * Sigma * w_mkt
        equilibrium = self.risk_aversion * np.dot(cov_matrix, market_weights)
        
        self._equilibrium_returns = equilibrium
        return equilibrium
    
    def _build_view_matrices(self, views: List[InvestorView]) -> Tuple[np.ndarray, np.ndarray]:
        """
        Build the P (pick) and Q (view) matrices from investor views.
        
        Returns:
            P: K x N matrix mapping assets to views
            Q: K x 1 vector of view expected returns
        """
        n_views = len(views)
        P = np.zeros((n_views, self.n_assets))
        Q = np.zeros(n_views)
        
        for k, view in enumerate(views):
            if view.view_type == "absolute":
                # Absolute view: single asset expected return
                for asset in view.assets:
                    if asset in self.asset_ids:
                        idx = self.asset_ids.index(asset)
                        P[k, idx] = 1.0
                Q[k] = view.expected_return
            
            elif view.view_type == "relative":
                # Relative view: outperformance between assets
                if len(view.assets) >= 2:
                    long_asset = view.assets[0]
                    short_asset = view.assets[1]
                    
                    if long_asset in self.asset_ids:
                        P[k, self.asset_ids.index(long_asset)] = 1.0
                    if short_asset in self.asset_ids:
                        P[k, self.asset_ids.index(short_asset)] = -1.0
                    
                    Q[k] = view.expected_return
        
        return P, Q
    
    def _build_omega(self, P: np.ndarray, cov_matrix: np.ndarray,
                     confidences: np.ndarray) -> np.ndarray:
        """
        Build the Omega matrix representing uncertainty in views.
        
        Omega is diagonal with elements proportional to view variance.
        Higher confidence means lower uncertainty.
        """
        n_views = P.shape[0]
        omega = np.zeros((n_views, n_views))
        
        for k in range(n_views):
            # View variance: confidence-weighted projection of covariance
            p_k = P[k:k+1, :].T
            view_variance = np.dot(p_k.T, np.dot(cov_matrix, p_k))[0, 0]
            
            # Scale by confidence (higher confidence = lower uncertainty)
            confidence = confidences[k]
            uncertainty_scale = self.tau * view_variance * (1 - confidence + 0.01)
            
            omega[k, k] = max(uncertainty_scale, 1e-8)
        
        return omega
    
    def compute_bl_returns(self, market_caps: np.ndarray, cov_matrix: np.ndarray,
                           views: List[InvestorView]) -> np.ndarray:
        """
        Compute Black-Litterman posterior expected returns.
        
        Args:
            market_caps: Market capitalizations for each asset
            cov_matrix: Covariance matrix of asset returns
            views: List of investor views with confidence levels
            
        Returns:
            Posterior expected returns vector
        """
        if cov_matrix.shape != (self.n_assets, self.n_assets):
            raise ValueError("Covariance matrix shape mismatch")
        
        if len(market_caps) != self.n_assets:
            raise ValueError("Market caps length mismatch")
        
        # Compute equilibrium returns
        pi = self._compute_equilibrium_returns(market_caps, cov_matrix)
        
        if not views:
            return pi.copy()
        
        # Build view matrices
        P, Q = self._build_view_matrices(views)
        confidences = np.array([v.confidence for v in views])
        
        # Build Omega (view uncertainty)
        Omega = self._build_omega(P, cov_matrix, confidences)
        
        # Black-Litterman formula:
        # E[R] = [(tau * Sigma)^-1 + P' * Omega^-1 * P]^-1 * 
        #        [(tau * Sigma)^-1 * Pi + P' * Omega^-1 * Q]
        
        tau_sigma_inv = np.linalg.inv(self.tau * cov_matrix + 1e-10 * np.eye(self.n_assets))
        omega_inv = np.linalg.inv(Omega + 1e-10 * np.eye(len(views)))
        
        # Posterior precision
        posterior_precision = tau_sigma_inv + np.dot(P.T, np.dot(omega_inv, P))
        
        # Posterior mean
        posterior_mean = np.dot(
            np.linalg.inv(posterior_precision + 1e-10 * np.eye(self.n_assets)),
            np.dot(tau_sigma_inv, pi) + np.dot(P.T, np.dot(omega_inv, Q))
        )
        
        return posterior_mean
    
    def optimize(self, market_caps: np.ndarray, cov_matrix: np.ndarray,
                 views: List[InvestorView], target_vol: Optional[float] = None) -> Dict[str, float]:
        """
        Compute optimal Black-Litterman weights.
        
        Args:
            market_caps: Market capitalizations
            cov_matrix: Covariance matrix
            views: Investor views
            target_vol: Optional target volatility for scaling
            
        Returns:
            Dictionary mapping asset IDs to weights
        """
        # Get posterior expected returns
        bl_returns = self.compute_bl_returns(market_caps, cov_matrix, views)
        
        # Mean-variance optimization: w = (1/delta) * Sigma^-1 * mu
        cov_inv = np.linalg.inv(cov_matrix + 1e-10 * np.eye(self.n_assets))
        raw_weights = np.dot(cov_inv, bl_returns) / self.risk_aversion
        
        # Normalize to sum to 1
        weights = raw_weights / (np.sum(raw_weights) + 1e-10)
        
        # Scale to target volatility if specified
        if target_vol is not None:
            port_vol = np.sqrt(np.dot(weights, np.dot(cov_matrix, weights)))
            if port_vol > 1e-10:
                weights *= target_vol / port_vol
        
        # Clip extreme weights
        weights = np.clip(weights, -0.5, 1.0)
        weights = weights / (np.sum(weights) + 1e-10)
        
        return dict(zip(self.asset_ids, weights))
    
    def get_nautilus_allocation(self, market_caps: np.ndarray, cov_matrix: np.ndarray,
                                 views: List[InvestorView], 
                                 portfolio_value: float) -> List[Dict]:
        """
        Generate Nautilus-compatible allocation commands.
        
        Returns:
            List of allocation dictionaries for Nautilus portfolio manager
        """
        weights = self.optimize(market_caps, cov_matrix, views)
        
        allocations = []
        for asset_id, weight in weights.items():
            if abs(weight) > 1e-6:
                allocations.append({
                    "instrument_id": asset_id,
                    "target_weight": float(weight),
                    "target_value": float(weight * portfolio_value),
                    "rebalance_threshold": 0.03,  # 3% drift tolerance
                    "source": "black_litterman"
                })
        
        return allocations
    
    def blend_with_hrp(self, hrp_weights: Dict[str, float], bl_weights: Dict[str, float],
                       bl_confidence: float) -> Dict[str, float]:
        """
        Blend Black-Litterman weights with HRP weights based on confidence.
        
        Args:
            hrp_weights: Weights from HRP optimizer
            bl_weights: Weights from Black-Litterman optimizer
            bl_confidence: Overall confidence in BL views (0 to 1)
            
        Returns:
            Blended weight dictionary
        """
        blended = {}
        
        for asset_id in self.asset_ids:
            hrp_w = hrp_weights.get(asset_id, 0.0)
            bl_w = bl_weights.get(asset_id, 0.0)
            
            # Confidence-weighted blend
            blended[asset_id] = (1 - bl_confidence) * hrp_w + bl_confidence * bl_w
        
        # Renormalize
        total = sum(blended.values())
        if abs(total) > 1e-10:
            blended = {k: v / total for k, v in blended.items()}
        
        return blended


def parse_views_from_soul(soul_data: Dict) -> List[InvestorView]:
    """
    Parse investor views from SOUL.md confidence scores.
    
    Args:
        soul_data: Dictionary containing ML ensemble predictions and confidence scores
        
    Returns:
        List of InvestorView objects
    """
    views = []
    
    # Extract alpha signals and confidence from SOUL data
    alpha_signals = soul_data.get("alpha_signals", {})
    confidence_scores = soul_data.get("confidence_scores", {})
    
    for asset, signal in alpha_signals.items():
        confidence = confidence_scores.get(asset, 0.5)
        
        # Only create views for high-confidence signals
        if confidence > 0.6:
            # Convert signal to expected return (scaled by confidence)
            expected_return = signal * confidence * 0.02  # 2% max daily
            
            views.append(InvestorView(
                assets=[asset],
                expected_return=expected_return,
                confidence=confidence,
                view_type="absolute"
            ))
    
    # Add relative views for pairs with divergent signals
    assets_list = list(alpha_signals.keys())
    for i, asset1 in enumerate(assets_list):
        for asset2 in assets_list[i+1:]:
            sig1 = alpha_signals.get(asset1, 0)
            sig2 = alpha_signals.get(asset2, 0)
            
            # Significant divergence
            if abs(sig1 - sig2) > 0.3:
                conf1 = confidence_scores.get(asset1, 0.5)
                conf2 = confidence_scores.get(asset2, 0.5)
                combined_conf = (conf1 + conf2) / 2
                
                if combined_conf > 0.5:
                    views.append(InvestorView(
                        assets=[asset1, asset2],
                        expected_return=(sig1 - sig2) * 0.01,
                        confidence=combined_conf,
                        view_type="relative"
                    ))
    
    return views


if __name__ == "__main__":
    # Example usage
    assets = ["BTC", "ETH", "SOL"]
    
    # Market capitalizations (in billions)
    market_caps = np.array([500, 250, 50])
    
    # Covariance matrix (annualized)
    cov_matrix = np.array([
        [0.04, 0.02, 0.015],
        [0.02, 0.06, 0.025],
        [0.015, 0.025, 0.09]
    ])
    
    # Sample views from ML ensemble
    views = [
        InvestorView(assets=["BTC"], expected_return=0.05, confidence=0.7),
        InvestorView(assets=["ETH"], expected_return=0.08, confidence=0.8),
        InvestorView(assets=["ETH", "SOL"], expected_return=0.03, confidence=0.6, view_type="relative")
    ]
    
    optimizer = BlackLittermanOptimizer(assets)
    weights = optimizer.optimize(market_caps, cov_matrix, views, target_vol=0.20)
    
    print("Black-Litterman Weights:")
    for asset, weight in weights.items():
        print(f"  {asset}: {weight:.4f}")
    
    allocations = optimizer.get_nautilus_allocation(
        market_caps, cov_matrix, views, portfolio_value=100000
    )
    print("\nNautilus Allocations:")
    for alloc in allocations:
        print(f"  {alloc}")
